pkg/exposer/pod_volume.go

```diff
@@ -0,0 +1,398 @@
+
+	// PeekExposed tests the status of the expose.
+	// If the expose is incomplete but not recoverable, it returns an error.
+	// Otherwise, it returns nil immediately.
+	PeekExposed(context.Context, corev1api.ObjectReference) error
+
+	// DiagnoseExpose generate the diagnostic info when the expose is not finished for a long time.
+	// If it finds any problem, it returns an string about the problem.
+	DiagnoseExpose(context.Context, corev1api.ObjectReference) string
+
+	// CleanUp cleans up any objects generated during the restore expose
+	CleanUp(context.Context, corev1api.ObjectReference)
+}
+
+// NewPodVolumeExposer creates a new instance of pod volume exposer
+func NewPodVolumeExposer(kubeClient kubernetes.Interface, log logrus.FieldLogger) PodVolumeExposer {
+	return &podVolumeExposer{
+		kubeClient: kubeClient,
+		fs:         filesystem.NewFileSystem(),
+		log:        log,
+	}
+}
+
+type podVolumeExposer struct {
+	kubeClient kubernetes.Interface
+	fs         filesystem.Interface
+	log        logrus.FieldLogger
+}
+
+var getPodVolumeHostPath = GetPodVolumeHostPath
+var extractPodVolumeHostPath = ExtractPodVolumeHostPath
+
+func (e *podVolumeExposer) Expose(ctx context.Context, ownerObject corev1api.ObjectReference, param PodVolumeExposeParam) error {
+	curLog := e.log.WithFields(logrus.Fields{
+		"owner":             ownerObject.Name,
+		"client pod":        param.ClientPodName,
+		"client pod volume": param.ClientPodVolume,
+		"client namespace":  param.ClientNamespace,
+		"type":              param.Type,
+	})
+
+	pod, err := e.kubeClient.CoreV1().Pods(param.ClientNamespace).Get(ctx, param.ClientPodName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "error getting client pod %s", param.ClientPodName)
+	}
+
+	if pod.Spec.NodeName == "" {
+		return errors.Errorf("client pod %s doesn't have a node name", pod.Name)
+	}
+
+	nodeOS, err := kube.GetNodeOS(ctx, pod.Spec.NodeName, e.kubeClient.CoreV1())
+	if err != nil {
+		return errors.Wrapf(err, "error getting OS for node %s", pod.Spec.NodeName)
+	}
+
+	curLog.Infof("Client pod is running in node %s, os %s", pod.Spec.NodeName, nodeOS)
+
+	path, err := getPodVolumeHostPath(ctx, pod, param.ClientPodVolume, e.kubeClient, e.fs, e.log)
+	if err != nil {
+		return errors.Wrapf(err, "error to get pod volume path")
+	}
+
+	path.ByPath, err = extractPodVolumeHostPath(ctx, path.ByPath, e.kubeClient, ownerObject.Namespace, nodeOS)
+	if err != nil {
+		return errors.Wrapf(err, "error to extract pod volume path")
+	}
+
+	curLog.WithField("path", path).Infof("Host path is retrieved for pod %s, volume %s", param.ClientPodName, param.ClientPodVolume)
+
+	hostingPod, err := e.createHostingPod(ctx, ownerObject, param.Type, path.ByPath, param.OperationTimeout, param.HostingPodLabels, param.HostingPodAnnotations, pod.Spec.NodeName, param.Resources, nodeOS)
+	if err != nil {
+		return errors.Wrapf(err, "error to create hosting pod")
+	}
+
+	defer func() {
```

Comment:
Seems the defer here is useless?

Reply:
Removed it. `createHostingPod` won't fail after the pod resource is created, so it is safe to remove it directly.